### PR TITLE
JDI: Workaround for opaque URL #330

### PR DESCRIPTION
**Pull request by jukzi**
prevents "IllegalArgumentException: URI is not hierarchical"

fixes https://github.com/eclipse-jdt/eclipse.jdt.debug/issues/330

### CODE REVIEW

File: org.eclipse.jdt.debug.tests/tests/org/eclipse/jdt/debug/tests/launching/LaunchTests.java [hunk 1]
```diff
@@ -41,6 +46,29 @@ public LaunchTests(String name) {
...
+			public String[] getSourceNames(String stratum) {
+				// opaque URL
+				return new String[] { "Main.java", "file:C:/workspace/prj/bi%20n/main/" };
+			}
+		};
+		File classesLocation2 = AdvancedSourceLookup.getClassesLocation(ref2);
+		assertNotNull(classesLocation1);
+		assertNotNull(classesLocation2);
+		assertEquals(classesLocation1, new File("C:/workspace/prj/bi n/main/"));
+		assertEquals(classesLocation2, new File("C:/workspace/prj/bi n/main/"));
```
[the remark this answers is not in the document]
Comment (jukzi) [reply org.eclipse.jdt.debug.tests/tests/org/eclipse/jdt/debug/tests/launching/LaunchTests.java]: well the "C:" is windows specific... does it work now?

File: org.eclipse.jdt.launching/launching/org/eclipse/jdt/internal/launching/sourcelookup/advanced/JDIHelpers.java [hunk 2]
```diff
@@ -75,10 +74,14 @@ public File getClassesLocation(Object element) throws DebugException {
 			try {
 				URL url = new URL(spec);
 				if ("file".equals(url.getProtocol())) { //$NON-NLS-1$
-					return new File(url.toURI()).toPath().normalize().toFile();
+					if (!spec.startsWith("file:/")) { //$NON-NLS-1$
+						// opaque -> URI is not hierarchical https://github.com/eclipse-jdt/eclipse.jdt.debug/issues/330
+						spec = spec.replace("file:", "file:/"); //$NON-NLS-1$ //$NON-NLS-2$
+					}
+					URI uri = URI.create(spec);
```
Comment (iloveeclipse) [inline org.eclipse.jdt.launching/launching/org/eclipse/jdt/internal/launching/sourcelookup/advanced/JDIHelpers.java:81]: `URL url = new URL(spec);` above is not really used. Should we just replace `if ("file".equals(url.getProtocol()))` with `if(spec.startsWith("file:")`, as we only use `spec` to create URI now?

If yes, we also should first trim the string like `spec = spec.trim();`

Comment (iloveeclipse) [reply org.eclipse.jdt.launching/launching/org/eclipse/jdt/internal/launching/sourcelookup/advanced/JDIHelpers.java:81]: Or will `url.getProtocol()` also answer "file" for absolute paths like `/hello/my/file` ?

Comment (jukzi) [reply org.eclipse.jdt.launching/launching/org/eclipse/jdt/internal/launching/sourcelookup/advanced/JDIHelpers.java:81]: that gives java.net.MalformedURLException: no protocol

**Pull request merged**